x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/ranges/ranges.test.tsx

```diff
@@ -0,0 +1,256 @@
+import { IndexPatternPrivateState, IndexPattern } from '../../../types';
+import { dataPluginMock } from '../../../../../../../../src/plugins/data/public/mocks';
+import { rangeOperation } from '../index';
+import { MODES, DEFAULT_INTERVAL, RangeIndexPatternColumn } from './ranges';
+import { autoInterval } from 'src/plugins/data/common';
+
+const defaultOptions = {
+  storage: {} as IStorageWrapper,
+  uiSettings: {} as IUiSettingsClient,
+  savedObjectsClient: {} as SavedObjectsClientContract,
+  dateRange: {
+    fromDate: 'now-1y',
+    toDate: 'now',
+  },
+  data: dataPluginMock.createStartContract(),
+  http: {} as HttpSetup,
+};
+
+describe('ranges', () => {
+  let state: IndexPatternPrivateState;
+  const InlineOptions = rangeOperation.paramEditor!;
+  const sourceField = 'MyField';
+
+  function setToHistogramMode() {
+    const column = state.layers.first.columns.col1 as RangeIndexPatternColumn;
+    column.dataType = 'number';
+    column.scale = 'interval';
+    column.params.type = MODES.Histogram;
+  }
+
+  function setToRangeMode() {
+    const column = state.layers.first.columns.col1 as RangeIndexPatternColumn;
+    column.dataType = 'string';
+    column.scale = 'ordinal';
+    column.params.type = MODES.Range;
+  }
+
+  beforeEach(() => {
+    state = {
+      indexPatternRefs: [],
+      indexPatterns: {},
+      existingFields: {},
+      currentIndexPatternId: '1',
+      isFirstExistenceFetch: false,
+      layers: {
+        first: {
+          indexPatternId: '1',
+          columnOrder: ['col1', 'col2'],
+          columns: {
+            // Start with the histogram type
+            col1: {
+              label: sourceField,
+              dataType: 'number',
+              operationType: 'range',
+              scale: 'interval',
+              isBucketed: true,
+              sourceField,
+              params: {
+                type: MODES.Histogram,
+                interval: autoInterval,
+                ranges: [{ from: 0, to: DEFAULT_INTERVAL, label: '' }],
+                maxBars: 1,
+              },
+            },
+            col2: {
+              label: 'Count',
+              dataType: 'number',
+              isBucketed: false,
+              sourceField: 'Records',
+              operationType: 'count',
+            },
+          },
+        },
+      },
+    };
+  });
+
+  describe('toEsAggConfig', () => {
+    afterAll(() => setToHistogramMode());
+
+    it('should reflect params correctly', () => {
+      const esAggsConfig = rangeOperation.toEsAggsConfig(
+        state.layers.first.columns.col1 as RangeIndexPatternColumn,
+        'col1',
+        {} as IndexPattern
+      );
+      expect(esAggsConfig).toEqual(
+        expect.objectContaining({
+          type: MODES.Histogram,
+          params: expect.objectContaining({
+            field: sourceField,
+            ranges: [{ from: 0, to: DEFAULT_INTERVAL }],
+            interval: autoInterval,
+            maxBars: 1,
```

Comment:
Okay so I think this is the source of many of the problems with esaggs. You're sending some parameters here that don't make sense. The easiest way to understand this is by adding some logging to `src/plugins/data/public/search/expressions/esaggs.ts` and use the Visualize editor in all the modes you're trying to copy.

I just tried this, and what I found are the following 3 cases with one special case:

1. Custom ranges:

```
type: "range"
params: {
  field: "bytes"
  ranges: [
    {from: 0, to: 1000}
    {from: 1000, to: 2000}
  ]
}
```

2. Auto interval histogram with auto bars:

```
type: "histogram"
params: {
  extended_bounds: {min: "", max: ""}
  field: "bytes"
  has_extended_bounds: false
  interval: "auto"
  min_doc_count: false
}
```

With max bars defined:

```
type: "histogram"
params: {
  extended_bounds: {min: "", max: ""}
  field: "bytes"
  has_extended_bounds: false
  interval: "auto"
  min_doc_count: false
  maxBars: 50
}
```

3. Histogram with minimum interval:

```
type: "histogram"
params: {
  extended_bounds: {min: "", max: ""}
  field: "bytes"
  has_extended_bounds: false
  interval: 1000
  min_doc_count: false
}
```

Reply:
I think I covered them all now.